src/hypothesis/strategies.py

```
@@ -775,19 +778,126 @@ def seed_random(seed):
+       with Hypothesis strategies, including extras where possible.
+    2. If ``thing`` is from the :mod:`python:typing` module, return the
+       corresponding strategy (special logic).
+    3. If ``thing`` has one or more subtypes in the merged lookup, return
+       the union of the strategies for those types that are not subtypes of
+       other elements in the lookup.
+    4. Finally, if ``thing`` has type annotations for all required arguments,
+       it is resolved via :func:`~hypothesis.strategies.builds`.
+
+    """
+    from hypothesis.searchstrategy import types
+    # Look for a known concrete type or user-defined mapping
+    lookup = dict(types.type_strategy_mapping())
+    lookup.update(types.generic_type_strategy_mapping())
+    lookup.update(types._global_type_to_strategy_lookup)
+    if thing in lookup:
+        strategy = lookup[thing]
+        if not isinstance(lookup[thing], SearchStrategy):
+            strategy = lookup[thing](thing)
+        if strategy.is_empty:
+            raise ResolutionFailed(
+                'Error: %r resolved to an empty strategy' % (thing,))
+        return strategy
+    # I tried many more elegant checks, but `typing` tends to treat the type
+    # system as a loose guideline at best so they were all unreliable.
+    if getattr(thing, '__module__', None) == 'typing':
+        return types.from_typing_type(thing)
+    # If there's no exact match above, use similar subtype resolution logic
+    strategies = [  # pragma: no cover
+        v if isinstance(v, SearchStrategy) else v(thing)
+        for k, v in lookup.items()
+        if isclass(k) and isclass(thing) and types.try_issubclass(k, thing) and
+        sum(types.try_issubclass(k, T) for T in lookup) == 1
+    ]
+    empty = ', '.join(repr(s) for s in strategies if s.is_empty)
+    if empty:
+        raise ResolutionFailed(
+            'Could not resolve %s to a strategy; consider using '
+            'register_type_strategy' % empty)
```

- I agree we don't, but `v` might be a function written by a user who likes the composition properties of empty strategies.  This check then converts it into ResolutionFailed
- `typing.Union`, per comment about coverage above
- A type without inhabitants is reasonable, but what strategy do you expect `from_type` to return?  The problem with `nothing()` is that it gets composed away by almost any operation, so I prefer to raise an immediate error - users can register an overriding strategy for cases where this is wrong.